addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/WebCacheDeceptionScanRule.java

```diff
@@ -0,0 +1,246 @@
+import java.net.URL;
+import java.util.ArrayList;
+import org.apache.commons.httpclient.URI;
+import org.apache.commons.text.similarity.LevenshteinDistance;
+import org.apache.hc.client5.http.classic.methods.*;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
+import org.apache.hc.client5.http.impl.classic.HttpClients;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.AbstractAppPlugin;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Category;
+import org.parosproxy.paros.network.HttpMessage;
+
+/**
+ * Active scan rule which attempts Web Cache Deception attack.
+ *
+ * @author Nikhil (@bettercalln1ck)
+ */
+public class WebCacheDeceptionScanRule extends AbstractAppPlugin {
+
+    private static final int LEVENSHTEIN_THRESHOLD = 200;
+
+    private static final String MESSAGE_PREFIX = "ascanalpha.webCacheDeception.";
+    private static final Logger LOG = LogManager.getLogger(WebCacheDeceptionScanRule.class);
+
+    protected static final String[] TEST_EXTENSIONS = {
+        "css", "jpg", "js", "html", "gif", "png", "svg", "php", "txt", "pdf", "asp"
+    };
+
+    @Override
+    public int getId() {
+        return 40039;
+    }
+
+    @Override
+    public String getName() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "name");
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "desc");
+    }
+
+    @Override
+    public void scan() {
+        if (initialTest()) {
+            HttpMessage authorisedMessage = getNewMsg();
+            try {
+                sendAndReceive(authorisedMessage);
+                String originalResponse = authorisedMessage.getResponseBody().toString();
+                URI uri = authorisedMessage.getRequestHeader().getURI();
+                ArrayList<String> extensions = new ArrayList<>();
+                String path = uri.getPath();
+                // checks whether the page with appended path gets cached or not
+                for (String ext : TEST_EXTENSIONS) {
+                    String newPath = path + "/test." + ext;
+                    uri.setPath(newPath);
+                    authorisedMessage.getRequestHeader().setURI(uri);
+                    sendAndReceive(authorisedMessage);
+                    String cachedResponse = authorisedMessage.getResponseBody().toString();
+                    String unauthorisedCachedResponse =
+                            makeUnauthorisedRequest(
+                                    uri, authorisedMessage.getRequestHeader().getMethod());
+
+                    if (checkSimilarity(unauthorisedCachedResponse, cachedResponse)) {
+                        extensions.add(ext);
```

Comment:
If one is added is it worth checking others?

Reply:
It is to identify the other extensions that are also vulnerable and display them in alert.